examples/expect-extend/toBeWithinRange.ts

```diff
@@ -6,13 +6,13 @@
  */
 
 import {expect} from '@jest/globals';
-import type {RawMatcherFn} from 'expect';
+import type {ExpectationResult} from 'expect';
 
-const toBeWithinRange: RawMatcherFn = (
```

Comment:
Interestingly `RawMatcherFn` is indirectly export through `jest.extend(matchers: MatchersObject)`. The `MatchersObject` is shaped by `RawMatcherFn`, that’s why `this` is correctly typed here (it is not declared, but gets typed nicely):

https://github.com/facebook/jest/blob/faef0b4b7082df574a0e4423b86d468847360f17/packages/expect/__typetests__/expect.test.ts#L28-L35

Reply:
Pattern `const toBeWithinRange: RawMatcherFn = () => {}` is sort of type casting. It took time for me to figure that out, but using types this way makes `toBeWithinRange` have typings only from `RawMatcherFn`. `actual` and `expected` will become `any` (or `unknown`), but never `string`, `number` or anything else. I try to say that `RawMatcherFn` is very good type (see comment above), but it should not be used for casting, because it just erases users types. Hence, if one needs `this` to be typed:

```ts
import {ExpectationResult, MatcherState} from 'expect';

function toHaveContext(this: MatcherState, received: string): ExpectationResult {
  // implementation
};
```

Hard to explain. I will make a couple of screenshots.

Reply:
Quick screenshots. Using `RawMatcherFn` (top line) the `expected` gets cast to `any` (at the bottom):
 
<img width="757" alt="Screenshot 2022-02-13 at 10 56 28" src="https://user-images.githubusercontent.com/72159681/153746191-d54b8415-9d52-4082-bb28-5aa29c3f58ac.png">

Same code, but `RawMatcherFn` is removed. `floor`, `ceiling` are typed as expected (bottom), only Prettier is complaining (top):

<img width="757" alt="Screenshot 2022-02-13 at 10 57 00" src="https://user-images.githubusercontent.com/72159681/153746171-b16f8c3d-fffd-4c31-b777-4c9b117c5674.png">

And one more, if `any`s are replaced with `unknown`s inside `RawMatcherFn`. TS is complaining. Here is the case, which made me understand that `RawMatcherFn` shouldn’t be used this way:

<img width="757" alt="Screenshot 2022-02-13 at 10 58 35" src="https://user-images.githubusercontent.com/72159681/153746318-dda638e8-2603-4275-95e6-fdecd2f8e056.png">

Reply:
Cool, thanks for the thorough explanation.

Not sure how to best handle this. You want to type your matcher with specific types, so that users using the matcher gets correct types. However, in your actual matchers, the arguments _are_ `unknown`, and you need to validate them. People might ignore the type error (or just not use typescript). Capturing that type error would be nice. Not sure if possible, tho 🙂

Reply:
Ah.. Now I get what you have in mind. Made it work with `unknown`s inside `RawMatcherFn`. Good point, this is closer to real world usage. TS is not complaining, but `RawMatcherFn` is still erasing arguments:

<img width="757" alt="Screenshot 2022-02-13 at 11 18 49" src="https://user-images.githubusercontent.com/72159681/153746788-0010ea1d-84d7-457b-9d6e-6eb651c2d688.png">

Seems better without `RawMatcherFn`:

<img width="757" alt="Screenshot 2022-02-13 at 11 19 06" src="https://user-images.githubusercontent.com/72159681/153746794-6895f171-4347-4a4a-a5cf-642ad2e8a7dd.png">

Perhaps documentations could mention that `unknown` is recommended, because it will remind to validate values?

Reply:
Wait.. With `unknow`s inside `RawMatcherFn` all works how you described (see at the bottom, `string` is not allow):

<img width="757" alt="Screenshot 2022-02-13 at 11 30 41" src="https://user-images.githubusercontent.com/72159681/153747088-09328b50-c36d-43bb-875b-4945c238d65b.png">

Reply:
Hm.. Tricky indeed.

Reply:
Yeah, might not be possible to solve in a clean way 🙂 Might make sense for authors to cast their matcher so the args are `unknown`, but the exported type is still strict?

Reply:
Perhaps `MatcherFunction` (should it be renamed?) could take one more generic type arg and do this casting? I will try out.

Reply:
Exciting!

Reply:
Great. I will clean up this draft, move it to the code base and push to this branch. Interestingly: what I described as a problem in the beginning, became very nice feature. Thanks for good questions!

Reply:
awesome, looking forward to it! these changes are a great step towards making `@types/jest` use types from this repo 👍